extensions/wikia/AdEngine3/src/templates/fmr-rotator.js

```diff
@@ -0,0 +1,257 @@
+ ****************************************/
+
+/**
+ * Checks if minimum refresh time counter is over
+ *
+ * @returns {boolean}
+ */
+function enoughTimeSinceLastRefresh() {
+	return ((new Date()) - refreshInfo.lastRefreshTime) > refreshInfo.refreshDelay;
+}
+
+/**
+ * Checks UAP is loaded to disable FRM logic in this case
+ *
+ * @returns {boolean}
+ */
+function fanTakeoverLoaded() {
+	return universalAdPackage.isFanTakeoverLoaded();
+}
+
+/**
+ * Checks and returns page navigation bar height
+ *
+ * @returns {int}
+ */
+function getNavbarHeight() {
+	const navbar = document.getElementById('globalNavigation');
+
+	return navbar ? navbar.offsetHeight : 0;
+}
+
+/**
+ * Checks if slot can be refreshed or limit is reached
+ *
+ * @returns {boolean}
+ */
+function refreshLimitAvailable() {
+	const limitAvailable = btRec || (currentAdSlot && (currentAdSlot.getConfigProperty('repeat.index') < 20));
+
+	if (!limitAvailable) {
+		scrollListener.removeCallback(rotatorListener);
+	}
+
+	return limitAvailable;
+}
+
+/**
+ * Shows or hides recirculation element
+ *
+ * @param {boolean} visible
+ * @returns {void}
+ */
+function swapRecirculation(visible) {
+	recirculationElement.style.display = visible ? 'block' : 'none';
+	refreshInfo.adVisible = !visible;
+}
+
+/**
+ * Checks if user is in the right position to rotate FMR
+ *
+ * @returns {boolean}
+ */
+function userInRightPosition() {
+	return utils.isInViewport(recirculationElement) || (btRec && currentRecNode && utils.isInViewport(currentRecNode))
+		|| (currentAdSlot && currentAdSlot.getElement() && utils.isInViewport(currentAdSlot.getElement()));
+}
+
+/**
+ * Update module refresh information after rotation
+ *
+ * @returns {void}
+ */
+function updateAdRefreshInformation() {
+	refreshInfo.lastRefreshTime = new Date();
+	refreshInfo.refreshAdPosition = window.scrollY;
+}
+
+/**
+ * Checks if user reached certain scroll position to enable FMR logic
+ *
+ * @returns {boolean}
+ */
+function userReachedStartPosition() {
+	return refreshInfo.startPosition <= window.scrollY;
+}
+
+/**
+ * Checks if user scrolled some distance before slot reload
+ *
+ * @returns {boolean}
+ */
+function userScrolledEnoughDistance() {
+	return Math.abs(window.scrollY - refreshInfo.refreshAdPosition) > refreshInfo.heightScrolledThreshold;
+}
+
+/**********************************
+ * Watchers and exported functions
+ **********************************/
+
+/**
+ * Rotates floating medrec based on given conditions
+ *
+ * @returns {void}
+ */
+function rotateSlots() {
+	if (enoughTimeSinceLastRefresh() && refreshLimitAvailable() && userScrolledEnoughDistance() && userInRightPosition()) {
+		if (refreshInfo.adVisible) {
+			if (btRec) {
+				removeRecNode();
+			} else {
+				slotTweaker.hide(currentAdSlot);
+			}
+
+			updateAdRefreshInformation();
+			swapRecirculation(true);
+		} else {
+			scrollListener.removeCallback(rotatorListener);
+
+			context.push('state.adStack', { id: currentSlotName });
```

Comment:
Should't SlotRepeater take care of that?

Reply:
Not really. SlotRepeater is not too clever and works only onScroll and we don't inject copies of ICB below but on the same viewport height. We need to move some fmr-rotate logic into it and I don't wanna do this, so I decided to push slots to ads queue manually.

Reply:
SlotRepeter repeats the slot. 

We can make it more clever, refactor the code. In my opinion fmr-rotator should only pass the conditions (promises?) when slot is going to be repeated, but repeating logic should be included somewhere in the ad-engine. 

I can see that it's going to end with rewriting whole code here, so let's leave it for now. But I believe it should be moved to ad-engine when it there is a second use case.

Reply:
👍